### PR TITLE
[Feature]: Improve index.html

### LINKED ISSUE #32: [Feature]: Improve index.html
### Feature Description

Create a better, static page experience based on the events we have. For instance, it would be nice to have:

- Today's MHN event
- Intro and link to the repo\
- Link to the ICS subscription file

Inline CSS and JS will probably work fine. We'll likely need some logic to parse the `events.json` file to figure out the current date. Then we'd need to parse that event from the JSON and show it in a timezone agnostic way (since an event from 1-4pm happens at 1-4pm regardless of timezone).

That means the only important bit of information is the current date which the browser will provide.

### PR DESCRIPTION
**Pull request by y-ashaswini**
## Issue

Relevant [issue number](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- Resolves Issue #32
- Suggested changes have been implemented

## Running Steps

1. npm run start (fetching data)
2. npm run server (starts a simple node [nodemon] server instance on port 8080)
3. can view the static site @ localhost:8080 or where it is hosted @ github pages (since it's all happening through index.html)

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1,11 +1,144 @@
...
+            <a href="https://github.com/seriouslysean/monster-hunter-now-events" target="_blank">Link</a>
+        </span>
+        <span>Download the ICS Subscription File:
+            <a href="https://seriouslysean.github.io/monster-hunter-now-events/dist/events.ics" target="_blank">Link</a>
+        </span>
+    </div>
 </body>
-</html>
+
+</html>
```
Comment (seriouslysean) [inline index.html:157]: Make sure all file end in a trailing return.

File: package.json [hunk 2]
```diff
@@ -66,7 +67,8 @@
     "test:logger": "ts-node --esm -r dotenv/config ./tools/test-logger.ts",
     "ts-node": "ts-node",
     "tsc": "tsc",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "server": "nodemon server.js"
```
Comment (seriouslysean) [inline package.json]: Switch to serve

File: server.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import http from 'http';
+import url from 'url';
+import fs from 'fs';
+
+const PORT = 8080;
```
Comment (seriouslysean) [inline server.js]: Add port as a value to the env, default to 8080 and use `process.env.PORT || 8080` to run this so it’s configurable. Make sure to run the nodemon start script with Dotenv.

File: server.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const index = fs.readFileSync('index.html');
+const events = fs.readFileSync('./dist/events.json');
```
Comment (seriouslysean) [inline server.js]: You should be able to use the path config value in the src/utils/config folder. Update these to work similarly to how I do pathing elsewhere.

File: server.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            res.end(events);
+            break;
+        default:
+            res.end(index);
+            break;
+    }
+}
+
+const server = http.createServer(createServer);
+server.listen(PORT);
```
Comment (seriouslysean) [inline server.js]: Add a callback function to this.

```suggestion
server.listen(PORT, () => {
    // move start log in here
});
```

File: server.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+console.log('Server listeninng on port', PORT);
```
Comment (seriouslysean) [inline server.js]: This should go inside the listen callback to ensure it fires after the server is listening.

File: server.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import http from 'http';
+import url from 'url';
```
Comment (seriouslysean) [inline server.js]: You could destructure parse directly as it’s the only function you use.

File: server.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import fs from 'fs';
```
Comment (seriouslysean) [inline server.js]: Destructure readFileSync directly.

File: server.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import http from 'http';
```
Comment (seriouslysean) [inline server.js]: Destructure createServer.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import url from 'url';
+import fs from 'fs';
+
+const PORT = 8080;
+const index = fs.readFileSync('index.html');
+const events = fs.readFileSync('./dist/events.json');
+
+function createServer(req, res) {
```
Comment (seriouslysean) [inline server.js]: Change name to `initServer` so it’s less confusing and won’t conflict with the changed http import.

File: server.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    const path = url.parse(req.url).pathname;
+
+    res.writeHead(200);
+
+    switch (path) {
+        case '/json':
```
Comment (seriouslysean) [inline server.js]: This shouldn’t be Inferring events from the path. This server should host this entire root folder, and use the /dist/events.json file directly via index.html just like GitHub does for the pages implementation/deployment.

File: index.html [hunk 12]
```diff
@@ -1,11 +1,144 @@
...
+        bottom:0px;
+        margin:0px;
+        padding: 40px;
+        background: #222;
+        color: #F6F4E4;
+    ">
+        <span>Repository:
+            <a href="https://github.com/seriouslysean/monster-hunter-now-events" target="_blank">Link</a>
+        </span>
+        <span>Download the ICS Subscription File:
```
Comment (seriouslysean) [inline index.html]:
```
Subscribe to the calendar:
```
The intention here is to have people subscribe, not download the file directly. By subscribing directly to the ICS file, automatic updates will be possible.

File: index.html [hunk 13]
```diff
@@ -1,11 +1,144 @@
...
+        flex-direction:column;
+        gap:20px;
+        position:absolute;
+        bottom:0px;
+        margin:0px;
+        padding: 40px;
+        background: #222;
+        color: #F6F4E4;
+    ">
+        <span>Repository:
```
Comment (seriouslysean) [inline index.html]: This should be a p tag.

File: index.html (hunk 12 above)
Comment (seriouslysean) [inline index.html]: This should also be a p tag.

File: index.html [hunk 14]
```diff
@@ -1,11 +1,144 @@
...
+            .catch(function (err) {
+                console.log('error: ' + err);
+            });
+
+    </script>
+
+
+    <!-- Footer -->
+    <div style="
+        width:100%;
```
Comment (seriouslysean) [inline index.html]: Move all inline styles to classes and define the styles inside of this classes in a `<style>` block in this file.

File: index.html [hunk 15]
```diff
@@ -1,11 +1,144 @@
...
+                                ">${event.name}</div>
+                                <div style="
+                                    font-size: 16px;
+                                    font-weight: 400;
+                                ">${event.desc}</div>
+                            </div>`
+                    });
+                }
+            })
+            .catch(function (err) {
```
Comment (seriouslysean) [inline index.html]: Nice use of catch, it might be worth moving to async/await and making the error more useful.

File: index.html [hunk 16]
```diff
@@ -1,11 +1,144 @@
...
+                console.log('error: ' + err);
```
Comment (seriouslysean) [inline index.html]:
```suggestion
                console.error('Unable to load events!', err);
```

File: index.html [hunk 17]
```diff
@@ -1,11 +1,144 @@
...
+                                    font-size: 32px;
+                                    font-weight: 700;
+                                ">
+                            No events scheduled for today!
+                        </div>`;
+                } else {
+                    eventsTodayList.forEach((event) => {
+                        tag_display.innerHTML +=
+                            `<div style="
+                                width: 566px;
```
Comment (seriouslysean) [inline index.html]: Move all styles inline and classes these elements instead.

File: index.html [hunk 18]
```diff
@@ -1,11 +1,144 @@
...
+            .then(function (response) {
+                return response.json();
+            })
+            .then(function (data) {
+                eventsTodayList = appendData(data);
+                // console.log("events: ", eventsTodayList);
+                document.getElementById("tag_display_date").innerHTML = TODAY.getDate() + "/" + TODAY.getMonth() + "/" + TODAY.getFullYear();
+
+                let tag_display = document.getElementById("tag_display_events");
+                console.log(eventsTodayList.length);
```
Comment (seriouslysean) [inline index.html]: Remove this loaf as it’s for debugging only, it comment it out.

File: index.html [hunk 19]
```diff
@@ -1,11 +1,144 @@
...
+        fetch('/json')
+            .then(function (response) {
+                return response.json();
+            })
+            .then(function (data) {
+                eventsTodayList = appendData(data);
+                // console.log("events: ", eventsTodayList);
+                document.getElementById("tag_display_date").innerHTML = TODAY.getDate() + "/" + TODAY.getMonth() + "/" + TODAY.getFullYear();
+
+                let tag_display = document.getElementById("tag_display_events");
```
Comment (seriouslysean) [inline index.html]:
```suggestion
                let tagDisplay = document.getElementById("tag_display_events");
```
Prefer camel casing. Snake casing for variables and such is more typical in Python.

Comment (seriouslysean) [inline index.html]:
```suggestion
                let tag_display = document.getElementById("tag-display-events");
```
Prefer kebab case for css ids and classes. Though in this case you could shorten the ID to just `events`.

File: index.html [hunk 20]
```diff
@@ -1,11 +1,144 @@
...
+                        desc: event.description,
+                    });
+                }
+            });
+
+            return eventsTodayList;
+        }
+
+
+        fetch('/json')
```
Comment (seriouslysean) [inline index.html]:
```suggestion
        fetch('/dist/events.json')
```
Prefer the actual hosted dist file so we get the generated and deduped content.

File: index.html [hunk 21]
```diff
@@ -1,11 +1,144 @@
...
+
+    <script type="text/javascript">
+
+        let eventsTodayList = [];
+        const TODAY = new Date();
+
+        function appendData(data) {
+            data.events.forEach((event) => {
+                // console.log("event: ", event);
+                let flag = 0;
```
Comment (seriouslysean) [inline index.html]: You likely don’t need a flag. You should be able to loop through the events and if it occurs today push that event directly in to the new collection.

I see what you were going for though, and the logic still works. The flag feels like unnecessary complexity.

File: index.html [hunk 22]
```diff
@@ -1,11 +1,144 @@
 <!DOCTYPE html>
 <html lang="en">
+
 <head>
     <meta charset="UTF-8">
-    <meta http-equiv="refresh" content="0;url=https://github.com/seriouslysean/monster-hunter-now-events">
```
Comment (seriouslysean) [inline index.html:5]: With the removal of this redirect, there needs to be a link to the GitHub homepage on this page somewhere, maybe as a very slim, centered footer.

**Review by seriouslysean — CHANGES_REQUESTED** [review]
This looks awesome! I think it’s really close, just a lot of small comments and then this is good to merge. This is an astounding amount of work and I really appreciate it!

File: index.html (hunk 12 above)
Comment (y-ashaswini) [reply index.html:138]: Oh, subscribe? Correct me if I'm wrong, but by subscribe do you mean setting up some live updates system? Some mailer system perhaps? Do you have some suggestions in mind to implement this?

File: server.js (hunk 9 above)
Comment (y-ashaswini) [reply server.js]: Right, done.

File: server.js (hunk 6 above)
Comment (y-ashaswini) [reply server.js]: Right, fair. Will do.

File: server.js (hunk 10 above)
Comment (y-ashaswini) [reply server.js]: Right

File: index.html (hunk 20 above)
Comment (y-ashaswini) [reply index.html]: Actually, this route (/json) is defined in the server.js file, where the original file is hosted from /dist/events.json
So the generaetd content is directly accessed from the original location anyway...

File: index.html (hunk 12 above)
Comment (y-ashaswini) [reply index.html]: oh, right

File: server.js (hunk 3 above)
Comment (y-ashaswini) [reply server.js]: Done

File: package.json (hunk 2 above)
Comment (y-ashaswini) [reply package.json]: Done

File: index.html (hunk 22 above)
Comment (y-ashaswini) [reply index.html:5]: I'm confused, where is this file? I'd edited this out... and yes, I've added a footer with links

File: index.html (hunk 1 above)
Comment (y-ashaswini) [reply index.html:157]: HTML files ending in a return?

File: index.html (hunk 12 above)
Comment (y-ashaswini) [reply index.html]: Apparently the ICS file isn't directly subscribe-able through code. The user must click on the downloaded file to open it in outlook. And the file must be downloaded locally first...

**Comment by y-ashaswini** [comment]
@seriouslysean , there are just a few things that I need some clarification with:

- You asked for a return statement at the end of index.html, please clarify how I might add that
- The calendar file with the .ics extension will not automatically subscribe the user to it, as far as I know, the only way to do it is to have the user manually download the file and click on it on their file explorer/ downloads folder, to open up the calendar in Outlook or any similar app
- You mentioned routing being compatible with the config.ts (./src/utils/config.ts), I tried importing the constant 'paths' from that file and using it to configure routes within server.ts but it didn't work. I then set it up the way it is in my latest commit. Please have a look at it and suggest changes if any.

Thanks!

File: index.html (hunk 1 above)
Comment (seriouslysean) [reply index.html:157]: Whoops, misread the extension. You’re right, no end of line here.

File: index.html (hunk 20 above)
Comment (seriouslysean) [reply index.html]: Right, but it’s still an alias to the file which is not preferable imo. So whether you change it here or in the server, either or both should point to the actual json file with the go between of the aliased path.

File: server.ts [hunk 23]
```diff
@@ -0,0 +1,32 @@
+import { createServer } from 'http';
+import { parse } from 'url';
+import { readFileSync } from 'fs';
+// import { paths } from './src/utils/config';
```
Comment (seriouslysean) [inline server.ts]: If you can’t get this to work feel free to remove it for the hardcoded path instead. I can look in to it as a follow up.

File: server.ts [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+const port = process.env.PORT || 8080;
+// const eventsFile = join(paths['dist'], 'events.json');
+// const defaultFile = join(paths['root'], 'index.html');
+
+function initServer(req: any, res: any) {
+    const path = parse(req.url).pathname;
+    console.log('current path: ', path);
+
+    res.writeHead(200);
+    if (path == '/dist/events.json') {
```
Comment (seriouslysean) [inline server.ts:12]: I’m surprised it needs a direct check like this. Is the server not able to just infer it’s a file? Admittedly I mostly work with express that abstracts a lot more of the lower level stuff.

**Review by seriouslysean — CHANGES_REQUESTED** [review]
This is awesome, couple small things and this is good to go. Seriously fantastic work and a huge qol!

File: index.html (hunk 22 above)
Comment (seriouslysean) [reply index.html:5]: This file is pushed to the hosted version of this repo, served via GitHub pages. If you added a link it’s all good!

File: index.html (hunk 12 above)
Comment (seriouslysean) [reply index.html]: Actually an ics file can be downloaded as you say, or if hosted in a server, subscribed to via url. That way if changes are added the clients auto update which is the intention here.

- https://support.apple.com/guide/calendar/subscribe-to-calendars-icl1022/mac
- https://support.microsoft.com/en-us/office/import-or-subscribe-to-a-calendar-in-outlook-on-the-web-503ffaf6-7b86-44fe-8dd6-8099d95f38df

File: server.ts (hunk 24 above)
Comment (y-ashaswini) [reply server.ts:12]: You're right actually, though we'd need middlewares and a few more installations and dependencies if we used express, so I thought we could go with the most basic bare-bones requirements for this one !

File: index.html (hunk 1 above)
Comment (y-ashaswini) [reply index.html:157]: Alright!

**Comment by y-ashaswini** [comment]
Thank you! Your suggestions have been very insightful, truly enjoyed contributing to your repo :)

File: server.ts [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+        res.end(readFileSync('dist/events.json'));
+    } else {
+        res.end(readFileSync('./index.html'));
+    }
+
+    // if (path == '/dist/events.json') {
```
Comment (seriouslysean) [inline server.ts]: Remove the commented out lines.

File: server.ts [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+    //     res.end(readFileSync(eventsFile));
+    // } else {
+    //     res.end(readFileSync(defaultFile));
+    // }
+}
+
+const server = createServer(initServer);
+server.listen(port, () => {
+    console.log('Server listeninng on port', port);
+});
```
Comment (seriouslysean) [inline server.ts:22]: Add new line to end of file.

File: package.json [hunk 27]
```diff
@@ -1,72 +1,74 @@
...
+        "lint:ts": "eslint --ext .ts --ignore-path .gitignore src/*",
+        "lint:prettier": "prettier --check src",
+        "lint": "npm run lint:ts && npm run lint:prettier",
+        "start": "npm run fetch:all-articles && npm run generate:events-json && npm run generate:events-ics",
+        "fetch:all-articles": "ts-node --esm -r dotenv/config ./tools/fetch-all-articles.ts",
+        "fetch:article": "ts-node --esm -r dotenv/config ./tools/fetch-article.ts",
+        "test:logger": "ts-node --esm -r dotenv/config ./tools/test-logger.ts",
+        "ts-node": "ts-node",
+        "tsc": "tsc",
+        "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (seriouslysean) [inline package.json]: The spacing of this entire file changed, make sure you’re only changing the scripture and packaged and not the amount of spaces on each line.

File: index.html [hunk 28]
```diff
@@ -1,11 +1,161 @@
...
+
+
+    <!-- Footer Section -->
+    <div class="footer">
+        <p>Repository:
+            <a href="https://github.com/seriouslysean/monster-hunter-now-events" target="_blank">Link</a>
+        </p>
+        <p>Save the date in my Calendar:
+            <a href="/dist/events.ics">Link</a>
+            <!-- <a href="https://seriouslysean.github.io/monster-hunter-now-events/dist/events.ics" target="_blank">Link</a> -->
```
Comment (seriouslysean) [inline index.html]: Delete the commented out lines.

File: index.html [hunk 29]
```diff
@@ -1,11 +1,161 @@
...
+        if (eventsTodayList.length === 0) {
+            tagDisplay.innerHTML +=
+                `<div class="cardHeading">
+                            No events scheduled for today!
+                        </div>`;
+        } else {
+            eventsTodayList.forEach((event) => {
+                tagDisplay.innerHTML +=
+                    `<div class="card">
+                                <div class="cardHeading">${event.name}</div>
```
Comment (seriouslysean) [inline index.html]: Make sure classes are hyphened and not camel case.
```suggestion
                                <div class="card-heading">${event.name}</div>
```

File: index.html [hunk 30]
```diff
@@ -1,11 +1,161 @@
...
+            position: absolute;
+            bottom: 0px;
+            margin: 0px;
+            padding: 40px;
+            background: #222;
+            color: #F6F4E4;
+        }
+    </style>
 </head>
+
```
Comment (seriouslysean) [inline index.html:75]: Remove extra line.

**Review by seriouslysean — CHANGES_REQUESTED** [review]
Last changes, then I’ll merge!

**Review by seriouslysean — DISMISSED** [review]
Seriously awesome, thank you! You put a ton of work in to this and it’s going to be such a good improvement!

**Review by seriouslysean — APPROVED** [review]

**Pull request merged**